subsys/dult/motion_detector.c

```diff
@@ -0,0 +1,346 @@
+		 * a sound playing action is finished.
+		 */
+		__ASSERT_NO_MSG(sound_cb);
+		sound_cb->sound_start();
+		poll_state = (poll_state == MOTION_POLL_STATE_PASSIVE) ?
+			     MOTION_POLL_STATE_PASSIVE_SOUND_REQUESTED :
+			     MOTION_POLL_STATE_ACTIVE_SOUND_REQUESTED;
+		sound_count++;
+
+		if (sound_count >= SEPARATED_UT_MAX_SOUND_COUNT) {
+			LOG_DBG("Stopping the motion detector: %d sounds played",
+				SEPARATED_UT_MAX_SOUND_COUNT);
+
+			motion_detector_stop();
+			backoff_setup();
+		}
+	} else {
+		(void) k_work_reschedule(&motion_poll_work,
+					 (poll_state == MOTION_POLL_STATE_PASSIVE) ?
+					 SEPARATED_UT_SAMPLING_RATE1 : SEPARATED_UT_SAMPLING_RATE2);
+	}
+}
+
+static void motion_poll_duration_work_handle(struct k_work *work)
+{
+	LOG_DBG("Stopping the motion detector: active poll duration timeout");
+
+	motion_detector_stop();
+	backoff_setup();
+}
+
+static void motion_poll_work_handle(struct k_work *work)
+{
+	if (poll_state == MOTION_POLL_STATE_PASSIVE) {
+		LOG_DBG("Passive motion polling");
+
+		motion_poll_handle();
+	} else if (poll_state == MOTION_POLL_STATE_ACTIVE) {
+		LOG_DBG("Active motion polling");
+
+		motion_poll_handle();
+	} else {
+		__ASSERT(0, "Invalid motion polling state");
+	}
+}
+
+static void separated_mode_transition_handle(void)
+{
+	int err;
+	int ret;
+	uint16_t separated_ut_timeout_period_seed;
+	uint32_t separated_ut_timeout_period;
+
+	/* Calculate the positive randomized time factor. */
+	err = sys_csrand_get(&separated_ut_timeout_period_seed,
+			     sizeof(separated_ut_timeout_period_seed));
+	if (err) {
+		LOG_WRN("FMDN State: sys_csrand_get failed: %d", err);
+
+		sys_rand_get(&separated_ut_timeout_period_seed,
+			     sizeof(separated_ut_timeout_period_seed));
+	}
+
+	BUILD_ASSERT(SEPARATED_UT_TIMEOUT_PERIOD_DIFF < UINT16_MAX);
+
+	/* Convert the random part range from <0; UINT16_MAX> to
+	 * <SEPARATED_UT_TIMEOUT_PERIOD_MIN; SEPARATED_UT_TIMEOUT_PERIOD_MAX>
+	 */
+	separated_ut_timeout_period = SEPARATED_UT_TIMEOUT_PERIOD_DIFF;
+	separated_ut_timeout_period *= separated_ut_timeout_period_seed;
+	separated_ut_timeout_period /= UINT16_MAX;
+	separated_ut_timeout_period += SEPARATED_UT_TIMEOUT_PERIOD_MIN;
+
+	LOG_DBG("Starting the work for enabling the motion detector. "
+		"Randomized timeout set to: %" PRIu32 "minutes", separated_ut_timeout_period);
+	__ASSERT_NO_MSG(!k_work_delayable_is_pending(&motion_enable_work));
+	ret = k_work_schedule(&motion_enable_work, K_MINUTES(separated_ut_timeout_period));
+	__ASSERT_NO_MSG(ret == 1);
+}
+
+static void near_owner_mode_transition_handle(void)
+{
+	if (poll_state != MOTION_POLL_STATE_STOPPED) {
+		LOG_DBG("Stopping the motion detector: owner nearby");
+		motion_detector_stop();
+	} else {
+		LOG_DBG("Motion detector is not running: owner nearby");
+	}
+	state_reset();
+}
+
+void dult_motion_detector_near_owner_state_change_notify(enum dult_near_owner_state_mode mode)
+{
+	switch (mode) {
```

Comment:
should we allow this state transition when the module is disabled? (dult_near_owner_state_set API can be called before dult_enable API)

Reply:
I will block it if the module is disabled and on module enable operation the module will check the state and act appropriately.